src/player.cpp

```diff
@@ -5116,6 +5095,139 @@ int player::hitall(int dam, int vary, Creature *source)
+    int part_num = -1;
+    vehicle *veh = g->m.veh_at( p, part_num );
+    if( critter != this && critter != nullptr ) {
+        target_name = critter->disp_name();
+        // Slamming into creatures and NPCs
+        // TODO: Handle spikes/horns and hard materials
+        armor_eff = 0.5f; // 2x as much as with the ground
+        // TODO: Modify based on something?
+        mod = 1.0f;
+        effective_force = force;
+    } else if( veh != nullptr ) {
+        // Slamming into vehicles
+        // TODO: Integrate it with vehicle collision function somehow
+        target_name = veh->disp_name();
+        if( veh->part_with_feature( part_num, "SHARP" ) != -1 ) {
+            // Now we're actually getting impaled
+            cut = force; // Lots of fun
+        }
+
+        mod = slam ? 1.0f : fall_damage_mod();
+        armor_eff = 0.25f; // Not much
+        if( !slam && veh->part_with_feature( part_num, "ROOF" ) ) {
+            // Roof offers better landing than frame or pavement
+            effective_force /= 2; // TODO: Make this not happen with heavy duty/plated roof
+        }
+    } else {
+        // Slamming into terrain/furniture
+        target_name = g->m.disp_name( p );
+        int hard_ground = g->m.has_flag( TFLAG_DIGGABLE, p ) ? 0 : 3;
+        armor_eff = 0.25f; // Not much
+        // Get cut by stuff
+        // This isn't impalement on metal wreckage, more like flying through a closed window
+        cut = g->m.has_flag( TFLAG_SHARP, p ) ? 5: 0;
+        effective_force = force + hard_ground;
+        mod = slam ? 1.0f : fall_damage_mod();
+        if( g->m.has_furn( p ) ) {
+            // TODO: Make furniture matter
+        } else if( g->m.has_flag( TFLAG_SWIMMABLE, p ) ) {
+            // TODO: Some formula of swimming
+            effective_force /= 4;
+        }
+    }
+
+    // Rescale for huge force
+    // At >30 force, proper landing is impossible and armor helps way less
+    if( effective_force > 30 ) {
+        // Armor simply helps way less
+        armor_eff *= 30.0f / effective_force;
+        if( mod < 1.0f ) {
+            // Landing helps only with the last 30 damage
+            const float scaled_damage = ( 30.0f * mod ) + effective_force - 30.0f;
+            mod = scaled_damage / effective_force;
+        }
+    }
+
+    int total_dealt = 0;
+    for( int i = 0; i < num_hp_parts; i++ ) {
+        const body_part bp = hp_to_bp( static_cast<hp_part>( i ) );
+        int bash = ( effective_force * rng(60, 100) / 100 );
+        damage_instance di;
+        di.add_damage( DT_BASH, bash, 0, armor_eff, mod );
+        // No good way to land on sharp stuff, so here modifier == 1.0f
+        di.add_damage( DT_CUT,  cut,  0, armor_eff, 1.0f );
+        total_dealt += deal_damage( nullptr, bp, di ).total_damage();
+    }
+
+    if( total_dealt > 0 ) {
+        // "You slam against the dirt" is fine
+        add_msg_if_player( m_bad, _("You are slammed against %s for %d damage."),
```

Comment:
Is it intended that you'll get no message if this happens to an NPC, because you'll get a message if the NPC is slammed against something, but has not taken any damage?